### ISSUE TITLE
Feature Request: Share list with other user

### ISSUE THREAD

**Issue by SYPHUNO** [karakeep-app/karakeep#37]
As described it would be awesome if we could share notes with other users on the server, for example share a list with a user and put the notes meant to be shared into said list.

This is a feature missing in usememos/memos and i think it would be great if you could implement it into Hoarder.

**Comment by MohamedBassem** [comment]
I can see this being useful indeed. I like it, will add it to the backlog!

**Comment by jjaychen1e** [comment]
@MohamedBassem How about share single bookmark with a link / ID, so I can reference it in my notes app

**Comment by MohamedBassem** [comment]
@SYPHUNO I've been thinking a bit more about this. When you say sharing a list, do you mean that the list is collaborative or is it a view only list?

Because if it's collaborative, I have a lot of questions about bookmark ownership in shared lists? For example, can you delete a bookmark that was added by someone else? Can you archive it? Those kind of question.

**Comment by SYPHUNO** [comment]
@MohamedBassem thanks for asking!
I drew up a little sketch, i hope it makes sense like this:
![grafik](https://github.com/MohamedBassem/hoarder-app/assets/80629577/8167be9f-c6aa-4a69-90b8-436a38b1dff2)
- Owner: Full control over the list, can add and remove users
- Edit: Can edit/remove/add posts
- Add: Can only add posts and maybe also edit posts added by themselves
- Read: Can only read, and see the list
(Now that i think of it, a dropdown might be better than boxes to check)

When it comes to archiving i'm not sure, maybe archiving a post would be user based, meaning:
If i archive a post it would become invisible in the original list for me, but others i've shared the list with with can still see it and archive it themselves?

**Comment by MohamedBassem** [comment]
Oh wow, thanks for taking the time and doing the sketch! That makes sense.

hmmm, I would like to keep it simple. So I'm thinking that maybe in shared lists, you don't get to modify the bookmark at all. You can either add a bookmark to that list, or remove it from the list. All other functionalities (e.g favouriting, archiving, deleting, etc) can't be done from the shared list view (they can still be done from the owner's homepage for example). What do you think?

**Comment by SYPHUNO** [comment]
No worries, i had fun and thought i would maybe portrait my idea better. :)

I totally understand your will to keep things simple, and think keeping feature such as archiving/deleting out of it sounds good.
If i understand your idea right, users can share a list between each other and add/remove bookmarks to this list from their personal homepage?
One thing i would really wish for would be the idea of collaborating in one bookmark or at least being able to edit bookmarks of other users in that list. 
Whats your opinion on the collaboration part?

**Comment by Markeyxs** [comment]
Just wanted to give this a +1 for the request. 
We understand that this feature is tricky to implement.
But sharing lists with the family without sharing one account would be so awesome. 
would even pay for this feature. 

Anyway keep up the good work :)

**Comment by bobonthenet** [comment]
I just started using Hoarder and the ability to share a list with other users would be huge for me. I would really like to be able to use this for planning an upcoming project with my wife.

**Comment by silmarine** [comment]
It would also be nice if we could share a read-only link to a list with someone that doesn't have an account. For example, I have a wishlist and want to send it to all my family and friends so they can see what I would like for christmas, but I don't want to make an account for each of them just for this function.

**Comment by MohamedBassem** [comment]
@silmarine this is tracked in the public lists issue (https://github.com/hoarder-app/hoarder/issues/126)

**Comment by WirtsLegs** [comment]
+1 for this, probably the one blocker keeping me from using Hoarder beyond just playing with it, Need to be able to have shared lists with my wife etc

**Comment by brainshower** [comment]
This is a key need here, and is making it difficult to wholeheartedly adopt Karakeep with my colleagues.  The use case is to designate certain lists that are "shared" with other select users (or all users) within my Karakeep install.  Anything added is just there for others to see.

This capability doesn't necessarily mean commenting would be required, or notifications.  Just initially, it can be designating lists that everyone with a Karakeep account can see and add/remove items.  Later you could add permissioning for users, and designate who is the "owner" vs. "viewer" and/or "editor"...but initially I would suggest going simple.

**Comment by luchfilip** [comment]
+1 for the lists. Not being able to share with family is a deal breaker.

**Comment by terri1769** [comment]
Would love to be able to publish a single list either to authenticated users or publicly.  The use case would be a wish list where I can bookmark wanted items from any and all vendors.  Notes on items would be nice for things like sizes or quantity wanted.

**Comment by LunaVoid** [comment]
Hi! Thought I might chime in and add onto what @terri1769 said. I currently use Hoarder/Karakeep as a convenient place to store academic papers and research that I am studying/want to build off of. It would be awesome if I could share this list publicly or privately with my research team without giving them access to all my other personal lists. It would be somewhat akin to a public pinterest board but much more versatile. 

Anyway thanks for this project, I'm really enjoying using it!

**Comment by ArturRuta** [comment]
Also chiming on this one. I think it would be appealing to share relevant readings with close friends or families. I see this would be of use for small scale communities. This is not X, reddit of facebook. It servers a different target audience, most likely moderate in size.

**Comment by Pytonballoon810** [comment]
+1

**Comment by ecvetek** [comment]
+1 The only reason why i didnt implement Karakeep at work for my colleagues yet. Hope to see this feature very soon.

**Comment by alphaecho35** [comment]
+1 This would be very useful!

**Comment by pnomme** [comment]
+1 Looking forward to this feature - frustrating having to run Linkwarden in parallel for my shared lists.

**Comment by landure** [comment]
I completely agree, sharing a list (and its nested lists) with users (and user groups) would me a very nice feature.
It would allow teams to build knowledge bases.

Linkwarden already allows sharing collections, but it requires sharing each nested collections independently.

**Comment by lexiismadd** [comment]
+1 on this!

**Comment by a-coruble** [comment]
Hello there, would also like to have this capability in Karakeep!

Does someone knows the development status towards making this a reality? I have some free time I'd like to invest working on this issue if possible.

From my understanding of @MohamedBassem comment, we should go first for a simple approach where a list can be shared by multiple people, who can only add / delete own entries to this shared list. Would that be the first sufficient milestone to consider this as resolved for the majority of folks here?

**Comment by MohamedBassem** [comment]
The main blocker for this issue is defining the semantics of shared lists. Some of the questions that I asked before but we didn't settle on an answer for is:

1. When an item is added to a shared list, do you expect it to appear in your homepage?
2. Marking an item that you don't own as "archived" probably doesn't have a meaning if it doesn't appear already on your homepage? So should we disallow it?
3. Can you favorite an item in a shared list that you don't own? I assume if you can, you want to also see it in your favorites list? What happens if the original owner removes it from the list afterwards? Do you expect a copy to be left in your account? If you do, then what else triggers a copy?
4. Can you see the notes that someone wrote to a bookmark?

Karakeep's current ownership/privacy model is what complicates the implementation of this issue. Today, a bookmark is owned by one and only one user. And the actions (favorite, archive, notes) are attached to the bookmark itself. I can use help in defining the semantics that you expect.

Another route is to consider list sharing as part of some federation feature and allow lists to be shared across servers as well. But that's a route I don't know much about and will require more research.

**Comment by a-coruble** [comment]
To answer your questions, for me, with my limited knowledge of the codebase:

1. Not a requirement / expectations for a first working version of this feature, I just want a place where me & my friends would know to share links relevant to one common topic.

2. Yup, I think we should, archiving might only be done by the owner of the bookmark, and shouldn't impact the shared list (i.e.: the bookmark should still be visible in the shared list).

3. Ideally, I would go with the behaviour you're describing where "favouring" a bookmark you don't own from a shared list would create a copy, since I could then use this copy as a "back-up". It would not add it back to the list automatically after copy. If the original owner deletes it afterwards, it should be deleted as usual, since we haven't linked back the copied version. It would then be to one of the member of the list with a copy of the bookmark to add it back if needed / wanted. Thinking about it, we would probably want to mark both the original from User A as favorite, as well as the new copy of User B, as not marking A's version would let users re-favorite and create duplicate entries of the same bookmark (or maybe it's not something we want to handle).
Though, for a first version, maye we could disallow favoring and any other actions beside adding / removing owned bookmarks from the list.

4. Could be interesting, I would say yes, without the ability of editing the notes.


After re-reading, I think that I would want Karakeep to have a "workspace" system where the bookmarks are linked to a workspace and not a person, and by default each user has it's own workspace. If you want to collaborate with friends or colleagues, you then create a workspace, where each list is shared, and where permissions would be handled from the workspace level. But this goes beyond the scope of the current discussion on sharing a list with people.

**Comment by landure** [comment]
Hi, to add my vision of thing:

- sharing a list can have this goals:
  1. create a full shared knowledge base, where any members can add, or remove, links (useful for teams).
  2. create a read-only news review which added posts appears in a dedicated "what's new" section of the homepage 
  3. create an add-only shared links base, that allows other users to find links the user found, and where links can only be deleted by the user having added the link, or the shared list administrators.

To have theses possibilities, the lists must have a quite complex permissions system. In my opinion, a quick implementation of the first option with public notes and private favorites / archive would be best.

Thank you for your work.

**Comment by amberflag48** [comment]
+1 for me as well, I see it as having multiple members to a list, since for me sharing would be done with other Karakeep users on the same server.

I would expect the list to show up to other members as well with an sharing icon so it can be differentiated from private lists.

Everything else would work the same way a normal list does.

Thanks for making this possible!

**Comment by ludwig234** [comment]
It would be really cool if you could use groups from OIDC or synced using LDAPs and then assign lists to groups.
Even better if you could specify permissions like "owner", "editor", and "viewer". 
That way a user group could own their own lists and share the lists with other individuals or groups as they seem fit.

**Comment by amuhareb** [comment]
voicing support for this as well!

**Comment by krom** [comment]
Now I'm choosing between Linkwarden and Karakeep.
Karakeep looks better in terms of working with pages, but it clearly lacks collaboration with lists.

**Comment by RonarPS** [comment]
Will there be an update to this usefull feature request? this is the one "missing link", to get it working as a full google keep replacement for me 👍

**Comment by Ra72xx** [comment]
> Now I'm choosing between Linkwarden and Karakeep. Karakeep looks better in terms of working with pages, but it clearly lacks collaboration with lists.

That's exactly my situation. As long as no internal sharing is supported in Karakeep, I will have to stick to Linkwarden.

**Comment by BK787** [comment]
New user here. Just set it on my network. Wanted to add support to this functionality request, as it's the reason I installed it. I have a young family and my partner and I planned to use this for coordinating 'life' and sharing ideas. Seen cool ideas for our son? Put it in a list. Decorating a room in the home? Put it in a list. I don't tend to use these kinda 'Pinterest' things myself but as a collaboration tool, I think they can be great.

So, on a basic level; I'd like the functionality to make a lists shared. A Google Photos/Immich style user level, as Syphuno sketched, would be great. Or just a simple shared and 'everyone gets the same permissions' would work for me, for now.

Thanks for the app and keep up the great work!! 

_Edit, to answer the questions from @MohamedBassem:_
> The main blocker for this issue is defining the semantics of shared lists. Some of the questions that I asked before but we didn't settle on an answer for is:
> 
> 1. When an item is added to a shared list, do you expect it to appear in your homepage?
> 2. Marking an item that you don't own as "archived" probably doesn't have a meaning if it doesn't appear already on your homepage? So should we disallow it?
> 3. Can you favorite an item in a shared list that you don't own? I assume if you can, you want to also see it in your favorites list? What happens if the original owner removes it from the list afterwards? Do you expect a copy to be left in your account? If you do, then what else triggers a copy?
> 4. Can you see the notes that someone wrote to a bookmark?

To sum up, from my perspective (as a fairly basic user)
1. I expect it to appear for everyone in the same way, if it is shared and they are part of a shared list etc.
2. If someone else edits soemthing that wasn't put there by them, perhaps the system can notify the 'author', so they can sanity check the edit?
3. Everyone should be able to favourite/comment any item that has been shared. In my mind, these shared areas are like a fruit/confectionary bowl in the middle of a table. Everyone can see them and everyone can interact with them. The nicest chocolate is hidden in my own personla cupboard...
4. If shared, yes. As mentioned, anything that is shared with others, should be visible to them and any comments/likes/whatever should be visible to any others within that 'share group'.

**Comment by swinis** [comment]
As more and more comments coming in requesting this feature - and as it is really "THE" thing missing (at least as I see it, and obviously not just me...) , it would be cool to see this request leaving the funnel and starting of directly into implementation phase 🥇 

@MohamedBassem ? :)

**Comment by MohamedBassem** [comment]
Ok folks, I've just merged shared lists in #2146 (our top most requested feature). It should be in the nightly build in ~30mins. This was a huge change, and I'd appreciate if people can give it a try, and report issues they face. As always, please take a backup before jumping on the nightly track.

Known issues:
1. There's currently no approval process for collaborative lists, once you add the user, they'll see the list immediately. Planning to fix this. 
2. Collaborative lists appear as a top level list in the side bar (and you can't set a parent up for them). Planning to fix that as well.
3. On the mobile app, collaborative lists will appear as if they're normal lists. However, if you attempt to favourite/edit a bookmark that you don't own, you'll be served an error message. Still didn't land the mobile app changes.
4. Lists that a user is a `viewer` in, still show up in the `manage lists` dialog and if you attempt to add a bookmark to a list that you're a viewer in, you'll get a generic `something went wrong` error.

**Comment by MohamedBassem** [comment]
All the known issues that are mentioned in the previous comment have been addressed. The feature should now be complete. Planning to cut the next release hopefully soon. Did anyone give it a try? Any feedback?

**Comment by Ra72xx** [comment]
For me it works (first-time Karakeep user, can't compare to previous releases). Users are all authenticated via OIDC. Two things I noticed:

- If I share with a user, I have to type in the complete mail address. No list to select a user by name. Is this intended?
- User has always to accept the shared list via email notification. I think at least an admin user should be able to define shared lists without further user interaction.

**Comment by krom** [comment]
I haven't set up SMTP yet, how can I accept the invitation without using email?

**Comment by MohamedBassem** [comment]
* Typing the full email is by design yes for privacy. In self-hosted installations, usually the users all know each others, but this can't be generalized.
* You don't need to email to accept the invitation, you can find the invitations in the "All lists" tab. I'll probably add a notification icon in the sidebar when you have unaccepted invitations.

**Comment by krom** [comment]
Yeah, right, I didn't see that.
Everything works fine, thanks.

**Comment by Ra72xx** [comment]
> * Typing the full email is by design yes for privacy. In self-hosted installations, usually the users all know each others, but this can't be generalized.

That's a valid point, however I propose that this privacy feature can be turned off by an admin. In small scale private installations, it is quite awkward if you have to type in (and remember!) other user‘s mail addresses.

**Comment by Unambiguous** [comment]
> * You don't need to email to accept the invitation, you can find the invitations in the "All lists" tab. I'll probably add a notification icon in the sidebar when you have unaccepted invitations.

Please, add a notification badge for pending invitations. Some users were reporting that they could not find the invitations. I did not know as well to look for them in All lists. This comment helped me find them.

**Comment by MohamedBassem** [comment]
@Unambiguous this is already implemented in the nightly build and will be coming out in the next release.
